sql/core/src/test/scala/org/apache/spark/sql/jdbc/v2/JDBCV2JoinPushdownIntegrationSuiteBase.scala

```diff
@@ -0,0 +1,522 @@
+    val df2 = sql(s"SELECT id, id FROM $catalogAndNamespace.${caseConvert(joinTableName1)}")
+
+    val rows = withSQLConf(SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "false") {
+      df1.join(df2, "id").collect().toSeq
+    }
+
+    withSQLConf(SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "true") {
+      val joinDf = df1.join(df2, "id")
+
+      checkJoinPushed(
+        joinDf,
+        s"$catalogAndNamespace.${caseConvert(joinTableName1)}, " +
+          s"$catalogAndNamespace.${caseConvert(joinTableName1)}"
+      )
+      checkAnswer(joinDf, rows)
+    }
+  }
+
+  test("Test aggregate on top of 2-way self join") {
+    val sqlQuery = s"""
+      |SELECT min(a.id + b.id), min(a.id)
+      |FROM $catalogAndNamespace.${caseConvert(joinTableName1)} a
+      |JOIN $catalogAndNamespace.${caseConvert(joinTableName1)} b ON a.id = b.id + 1
+      |""".stripMargin
+
+    val rows = withSQLConf(SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "false") {
+      sql(sqlQuery).collect().toSeq
+    }
+
+    withSQLConf(SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "true") {
+      val df = sql(sqlQuery)
+
+      checkAggregateRemoved(df)
+      checkJoinPushed(
+        df,
+        s"$catalogAndNamespace.${caseConvert(joinTableName1)}, " +
+          s"$catalogAndNamespace.${caseConvert(joinTableName1)}"
+      )
+
+      checkAnswer(df, rows)
+    }
+  }
+
+  test("Test aggregate on top of multi-way self join") {
+    val sqlQuery = s"""
+      |SELECT min(a.id + b.id), min(a.id), min(c.id - 2)
+      |FROM $catalogAndNamespace.${caseConvert(joinTableName1)} a
+      |JOIN $catalogAndNamespace.${caseConvert(joinTableName1)} b ON b.id = a.id + 1
+      |JOIN $catalogAndNamespace.${caseConvert(joinTableName1)} c ON c.id = b.id - 1
+      |""".stripMargin
+
+    val rows = withSQLConf(SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "false") {
+      sql(sqlQuery).collect().toSeq
+    }
+
+    withSQLConf(SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "true") {
+      val df = sql(sqlQuery)
+
+      checkJoinPushed(
+        df,
+        s"$catalogAndNamespace.${caseConvert(joinTableName1)}," +
+          s" $catalogAndNamespace.${caseConvert(joinTableName1)}, " +
+          s"$catalogAndNamespace.${caseConvert(joinTableName1)}")
+      checkAnswer(df, rows)
+    }
+  }
+
+  test("Test sort limit on top of join is pushed down") {
+    val sqlQuery = s"""
+      |SELECT min(a.id + b.id), a.id, b.id
+      |FROM $catalogAndNamespace.${caseConvert(joinTableName1)} a
+      |JOIN $catalogAndNamespace.${caseConvert(joinTableName1)} b ON b.id = a.id + 1
+      |GROUP BY a.id, b.id
+      |ORDER BY a.id
+      |LIMIT 1
+      |""".stripMargin
+
+    val rows = withSQLConf(SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "false") {
+      sql(sqlQuery).collect().toSeq
+    }
+
+    withSQLConf(
+      SQLConf.DATA_SOURCE_V2_JOIN_PUSHDOWN.key -> "true") {
+      val df = sql(sqlQuery)
+
+      checkSortRemoved(df)
```

Comment:
This one depends on scan builder implementor, we should make overridable member which tells whether sortLimit is supported and provide here that value:
Like:
```suggestion
      checkSortRemoved(df, sortLimitPushdownSupported)
```

Reply:
added `supportsXYZpushdown` in DataSourcePushdownTestUtils. Suites that extend it can override these values if some of the pushdowns are not supported